aptos-move/move-examples/post_mint_reveal_nft/sources/whitelist.move

```diff
@@ -0,0 +1,145 @@
+        assert!(bucket_table::contains(&whitelist_stage.whitelisted_address, &minter_address), error::permission_denied(EACCOUNT_NOT_WHITELISTED));
+        let remaining_minting_amount = bucket_table::borrow_mut(&mut whitelist_stage.whitelisted_address, minter_address);
+        assert!(*remaining_minting_amount >= user_minting_amount, error::invalid_argument(EEXCEEDS_MINT_LIMIT));
+        *remaining_minting_amount = *remaining_minting_amount - user_minting_amount;
+    }
+
+    /// Adds a new whitelist stage.
+    public fun add_whitelist_stage(admin: &signer, whitelist_start_time: u64, whitelist_end_time: u64, whitelist_price: u64, whitelist_stage: u64) acquires WhitelistMintConfig {
+        assert!(whitelist_start_time < whitelist_end_time, error::invalid_argument(EINVALID_WHITELIST_SETTING));
+        if (!whitelist_config_exists(signer::address_of(admin))) {
+            init_whitelist_config(admin);
+        };
+        let num_stages = get_num_of_stages(signer::address_of(admin));
+        assert!(whitelist_stage <= num_stages, error::invalid_argument(EINVALID_STAGE));
+        let config = borrow_global_mut<WhitelistMintConfig>(signer::address_of(admin));
+
+        // If whitelist_stage equals num_stages, it means that the user wants to add a new stage at the end of the whitelist stages.
+        if (whitelist_stage == num_stages) {
+            let whitelist_stage = WhitelistStage {
+                whitelisted_address: bucket_table::new<address, u64>(4),
+                whitelist_mint_price: whitelist_price,
+                whitelist_minting_start_time: whitelist_start_time,
+                whitelist_minting_end_time: whitelist_end_time,
+            };
+            vector::push_back(&mut config.whitelist_configs, whitelist_stage);
+        } else {
+            let whitelist_stage_to_be_updated = vector::borrow_mut(&mut config.whitelist_configs, whitelist_stage);
+            whitelist_stage_to_be_updated.whitelist_mint_price = whitelist_price;
+            whitelist_stage_to_be_updated.whitelist_minting_start_time = whitelist_start_time;
+            whitelist_stage_to_be_updated.whitelist_minting_end_time = whitelist_end_time;
+        };
+    }
+
+    /// Adds addresses to a specified whitelist stage.
+    public fun add_whitelist_addresses(admin: &signer, wl_addresses: vector<address>, mint_limit: u64, whitelist_stage: u64) acquires WhitelistMintConfig {
+        let config = borrow_global_mut<WhitelistMintConfig>(signer::address_of(admin));
+        assert!(whitelist_stage < vector::length(&config.whitelist_configs), error::invalid_argument(EINVALID_STAGE));
+        let whitelist_stage = vector::borrow_mut(&mut config.whitelist_configs, whitelist_stage);
+        let now = timestamp::now_seconds();
+        assert!(now < whitelist_stage.whitelist_minting_end_time, error::invalid_argument(EINVALID_UPDATE_AFTER_MINTING));
+
+        vector::for_each_ref(&wl_addresses, |wl_address| {
```

Comment:
Sorry for the back and forth on this. You might need to revert back to using a while loop here since vector::for_each_ref might not be available in mainnet in time to deploy this contract. We can update this to use vector::for_each_ref later.